### ISSUE TITLE
GPU 

### ISSUE THREAD

**Issue by Ivanyan0516** [Jia-Research-Lab/PointGroup#34]
support model = torch.nn.DataParallel(model,device_ids=[0,1])? or how can I train on several gpus? please

**Comment by xiaodongww** [comment]
Same quesiton here. In my case, I use ` torch.nn.parallel.DistributedDataParallel` for multi-gpu training. 
```python
    device = torch.device(f'cuda:{cfg.local_rank}')
    model = torch.nn.parallel.DistributedDataParallel(model).to(device)
```

I tried to train the model with two GPU cards with the same parameter settings as the authors'. The problem is that the loss curve is always higher than training on a single GPU card. Here is the loss curve
![image](https://user-images.githubusercontent.com/33693065/104885889-c1268300-59a3-11eb-9f4f-6ac194490f1f.png)

I am trying to figure it out.

Also, I am very curious about how to speed up the training process without losing performance using multiple gpu cards. Any suggestion will be helpful. Thanks! @llijiang

**Comment by tpzou** [comment]
> support model = torch.nn.DataParallel(model,device_ids=[0,1])? or how can I train on several gpus? please

As the code merges the batch, it can't be train used **DataParallel**

**Comment by tpzou** [comment]
Hi, @xiaodongww . Would you mind send me an copy about your code or contact? 

 I train to use the **spawn** and **DistributedDataParallel** , but it seems something wrong with my code.

**Comment by llijiang** [comment]
Hi, 
I have added **DistributedDataParallel** in this repo. Please refer to the ddp branch. 
It could be normally trained but I haven't tested the performance yet.  To train it with ddp, for example, when using 2 gpus, set `ngpu_per_node` in config file as 2 (batch size 8), and run
```
CUDA_VISIBLE_DEVICES=0,1 python train_ddp.py --config config/pointgroup_run2_scannet.yaml
```
You may try different parameters (e.g., lr, epochs) in the training.

**Comment by Ivanyan0516** [comment]
thanks so much!发自我的iPhone------------------ Original ------------------From: Li Jiang  <notifications@github.com>Date: Tue,Jan 19,2021 10:15 PMTo: Jia-Research-Lab/PointGroup <PointGroup@noreply.github.com>Cc: Ivanyan0516 <Ivanyan0516@gmail.com>, Author <author@noreply.github.com>Subject: Re: [Jia-Research-Lab/PointGroup] GPU  (#34)
Hi,
I have added DistributedDataParallel in this repo. Please refer to the ddp branch.
It could be normally trained but I haven't tested the performance yet.  To train it with ddp, for example, when using 2 gpus, set ngpu_per_node in config file as 2 (batch size 8), and run
CUDA_VISIBLE_DEVICES=0,1 python train_ddp.py --config config/pointgroup_run2_scannet.yaml

You may try different parameters (e.g., lr, epochs) in the training.

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/Jia-Research-Lab/PointGroup/issues/34#issuecomment-762865756",
"url": "https://github.com/Jia-Research-Lab/PointGroup/issues/34#issuecomment-762865756",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by Guptajakala** [comment]
@Ivanyan0516 
Hi, have you tried training with the author's released DistributedDataParallel version? Were you able to get similar peformance to single gpu?

**Comment by Ivanyan0516** [comment]
> @Ivanyan0516
> Hi, have you tried training with the author's released DistributedDataParallel version? Were you able to get similar peformance to single gpu?

Hi，I got 0.571 AP _50,better than single gpu 0.567,but still worse than the expected result, have you tried?
